---
package: SpatialCPie

title: Spatial transcriptomics cluster analysis with SpatialCPie

author:
- name: Joseph Bergenstråhle
  affiliation:
    Science for Life Laboratory, Department of Gene Technology, School of
    Biotechnology, Royal Institute of Technology (KTH), SE-106 91 Solna, Sweden
- name: Ludvig Bergenstråhle
  affiliation:
    Science for Life Laboratory, Department of Gene Technology, School of
    Biotechnology, Royal Institute of Technology (KTH), SE-106 91 Solna, Sweden
- name: Joakim Lundeberg
  affiliation:
    Science for Life Laboratory, Department of Gene Technology, School of
    Biotechnology, Royal Institute of Technology (KTH), SE-106 91 Solna, Sweden

vignette: >
  %\VignetteIndexEntry{SpatialCPie}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}

output: BiocStyle::html_document
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
    collapse = TRUE,
    comment = "# >",
    dev = "jpeg"
)
```

Introduction
============

SpatialCPie is an R package designed to facilitate cluster evaluation for
Spatial Transcriptomics (ST) data by providing intuitive visualizations that
display the relationship between clusters in order to guide the user during
cluster identification, selection and further downstream applications

Usage example
=============

```{r, include = FALSE}
# Patch SpatialCPie::runCPie so that it returns a mock output object instead of
# starting the web server
assignInNamespace(
    ns = "SpatialCPie",
    x = "runCPie",
    value = function(
        counts,
        assignments,
        image = NULL,
        spotCoordinates = NULL,
        view = NULL
    ) {
        data <- SpatialCPie:::.preprocessData(
            counts,
            assignments,
            spotCoordinates
        )
        likenessScores <- lapply(
            tail(data$distances, -1),
            function(x) SpatialCPie:::.likeness(x, 1.0)
        )
        arrayPlots <- lapply(
            likenessScores,
            function(x) SpatialCPie:::.arrayPlot(
                scores = x, coordinates = data$coordinates,
                image =
                    if (!is.null(image) && !is.null(data$coordinates))
                        grid::rasterGrob(
                            image,
                            width  = grid::unit(1, "npc"),
                            height = grid::unit(1, "npc"),
                            interpolate = TRUE
                        )
                    else NULL
            ) +
             ggplot2::scale_fill_manual(values = data$colors)
        )
        list(
            clusters = do.call(cbind, data$assignments),
            treePlot = SpatialCPie:::.clusterTree(
                do.call(cbind, data$assignments),
                transitionLabels = TRUE
            ) +
             ggplot2::scale_color_manual(values = data$colors),
            piePlots = arrayPlots,
            piePlotsInfo = likenessScores
        )
    }
)
```

```{r}
library(SpatialCPie)
set.seed(42)
```

Input data
----------

In this example, we will use a downsampled dataset from an experiment on the
human heart.
We begin by loading the count data[^data_pipeline]:

```{r}
counts <- read.table(
    system.file("extdata", "counts.tsv", package = "SpatialCPie"),
    sep = "\t",
    check.names = FALSE
)
counts[1:5, 1:5]
```

[^data_pipeline]: Generated by the [ST
pipeline](https://github.com/SpatialTranscriptomicsResearch/st_pipeline)

To overlay the data on the tissue, we also need to load the tissue image and its
corresponding spot data[^spot_pipeline], which specifies the pixel coordinates
of each spot:

```{r}
tissue <- jpeg::readJPEG(
    system.file("extdata", "he_image.jpg", package = "SpatialCPie")
)
spots <- parseSpotFile(
    system.file("extdata", "spot_data.tsv", package = "SpatialCPie")
)
head(spots)
```

[^spot_pipeline]: Generated by the [ST spot
detector](https://github.com/SpatialTranscriptomicsResearch/st_spot_detector).

Preprocessing
-------------

Typically, it's good to conduct some data filtering prior to the analysis:
This could include removing spots that are outside of the tissue or removing
spots or genes that have a low number of reads.

Since the spot file only contains the spots that are under the tissue, we can
use it to subset the counts:

```{r}
counts <- counts[, which(colnames(counts) %in% rownames(spots))]
```

Let's also remove all spots and genes that have less than 20 reads in total:

```{r}
repeat {
    d <- dim(counts)
    counts <- counts[rowSums(counts) >= 20, colSums(counts) >= 20]
    if (all(dim(counts) == d)) {
        break
    }
}
```

Computing cluster assignments
-----------------------------

Apart from the count data, the primary input to the gadget is a list of *cluster
assignments*.
The cluster assignments are labels assigned to each spot over different *cluster
resolutions*, where we use the terminology "(cluster) resolution k" to refer to
a partitioning of the spots into $k$ clusters.

Any clustering algorithm can be used to compute the cluster assignments.
For example, using `kmeans` to compute resolutions 2--5, we can write:

```{r}
assignments <- lapply(2:4, function(x) kmeans(t(counts), centers = x)$cluster)
str(assignments)
```

Visualization
-------------

We are now ready to use the gadget to visualize the data:

```{r}
result <- runCPie(
    counts,
    assignments,
    image = tissue,
    spotCoordinates = spots
)
```

The gadget has two main elements: the **cluster tree** and the **spatial array
plots**.
The cluster tree is interactive and by selecting rows of nodes, the
corresponding spatial array plots are also displayed.
The information contained in these plots will be discussed in the following
sections.

To exit the gadget, press the "Done" button.
The plots in the gadget, the cluster assignments, and the likeness scores are
stored in the output object, which has the following structure:

```{r}
str(result, max.level = 1)
```

### Cluster tree

```{r}
result$treePlot
```

The cluster tree is a visual representation of how the spatial features
transition from clusters of lower resolution to clusters of higher resolution.
The edge opacities show the proportion of spots that transition to or from a
given cluster while node radii display the size of each cluster.

When launching the gadget, spots are relabeled so as to minimize the number of
crossovers (label switches) between resolutions in the data.
Moreover, cluster color labels are selected so that dissimilar clusters are
farther away in color space.

### Spatial array plots

```{r}
result$piePlots$`4`
```

The spatial array plots show the spatial locations of the clusters.
The pie diagrams are based on a **likeness score** between each spot and each
cluster.
The likeness score, $L$, between spot $s$ and cluster $k$ is defined as:

$$
L(s, k) = \mbox{exp}\left(
    -\lambda ||x_s - \mbox{mean}({\{x_{s'}\}_{s' \in C(k)}})||_2
\right),
$$

where $x_i$ is the gene expression vector of spot $i$, $C(k)$ is the set of
spots in cluster $k$, and $\lambda$ is a user-defined constant (adjustable from
the gadget).

Note that the definition implies that it is possible to (practically) hard-label
the spots by setting $\lambda$ to a high value.

Sub-clustering
--------------

After a "global" analysis of the entire tissue section, it is often interesting
to sub-cluster parts of the tissue.
This can be achieved via cluster selection in the gadget and subsequent
sub-sampling of the clusters that the user wants to re-cluster further.

As a practical example, from the array plot above, we can see that clusters 1
and 4 in resolution 4 appear in mostly the same spatial areas and, from the
cluster tree, we can also see that they are mostly derived from the same
lower-resolution cluster.
Therefore, it could be interesting to sub-cluster the spots assigned to these
clusters:

```{r}
subcluster <- names(result$clusters[, "4"][
    which(result$clusters[, "4"] %in% c(1, 4))]
)
head(subcluster)
```

```{r}
subclusterResult <- runCPie(
    counts = counts[, subcluster],
    assignments = lapply(
        2:4,
        function(x) kmeans(t(counts[, subcluster]), centers = x)$cluster
    ),
    image = tissue,
    spotCoordinates = spots
)
subclusterResult$treePlot
```

From the array plots, the sub-clustering appears to pick up two different
clusters of stromal cells.
However, the pie charts show that most spots have quite comparable likeness to
each of the two clusters, which could indicate that this level of clustering is
too fine-grained.

```{r}
subclusterResult$piePlots$`2`
```
