---
package: SpatialCPie

title: Spatial transcriptomics cluster analysis with SpatialCPie

author:
- name: Joseph Bergenstråhle
  affiliation:
    Science for Life Laboratory, Department of Gene Technology, School of
    Biotechnology, Royal Institute of Technology (KTH), SE-106 91 Solna, Sweden
- name: Ludvig Bergenstråhle
  affiliation:
    Science for Life Laboratory, Department of Gene Technology, School of
    Biotechnology, Royal Institute of Technology (KTH), SE-106 91 Solna, Sweden
- name: Joakim Lundeberg
  affiliation:
    Science for Life Laboratory, Department of Gene Technology, School of
    Biotechnology, Royal Institute of Technology (KTH), SE-106 91 Solna, Sweden

vignette: >
  %\VignetteIndexEntry{SpatialCPie}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}

output: BiocStyle::html_document
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "# >"
)
```

Introduction
============

SpatialCPie is an R package designed to facilitate cluster evaluation for
Spatial Transcriptomics (ST) data by providing intuitive visualizations that
display the relationship between clusters in order to guide the user during
cluster identification, selection and further downstream applications

Usage example
=============

```{r}
library(SpatialCPie)
set.seed(123)
```

Input data
----------

In this example, we will use a downsampled dataset from an experiment on the
human heart.
We begin by loading the count data[^data_pipeline]:

```{r}
counts <- read.table(
    system.file("extdata", "counts.tsv", package = "SpatialCPie"),
    sep = "\t",
    check.names = F
)
counts[1:5, 1:5]
```

[^data_pipeline]: Generated by the [ST
pipeline](https://github.com/SpatialTranscriptomicsResearch/st_pipeline)

To overlay the data on the tissue, we also need to load the tissue image and its
corresponding spot data[^spot_pipeline], which specifies the pixel coordinates
of each spot:

```{r}
tissue <- jpeg::readJPEG(
    system.file("extdata", "he_image.jpg", package = "SpatialCPie")
)
spots <- parseSpotFile(
    system.file("extdata", "spot_data.tsv", package = "SpatialCPie")
)
head(spots)
```

[^spot_pipeline]: Generated by the [ST spot
detector](https://github.com/SpatialTranscriptomicsResearch/st_spot_detector).

Preprocessing
-------------

Typically, it's good to conduct some data filtering prior to the analysis:
This could include removing spots that are outside of the tissue or removing
spots or genes that have a low number of reads.

Since the spot file only contains the spots that are under the tissue, we can
use it to subset the counts:

```{r}
counts <- counts[, which(colnames(counts) %in% rownames(spots))]
```

Let's also remove all spots and genes that have less than 20 reads in total:

```{r}
d <- dim(counts)
repeat {
  counts <- counts[rowSums(counts) >= 20, colSums(counts) >= 20]
  if (all(dim(counts) == d)) {
    break
  }
  d <- dim(counts)
}
```

Computing cluster assignments
-----------------------------

Apart from the count data, the primary input to the gadget is a list of *cluster
assignments*.
The cluster assignments are labels assigned to each spot over different *cluster
resolutions*, where we use the terminology "(cluster) resolution k" to refer to
a partitioning of the spots into $k$ clusters.

Any clustering algorithm can be used to compute the cluster assignments.
For example, using `kmeans` to compute resolutions 2--5, we can write:

```{r}
assignments <- lapply(2:5, function(x) kmeans(t(counts), centers = x)$cluster)
str(assignments)
```

Visualization
-------------

We are now ready to use the gadget to visualize the data:

```{r, eval = FALSE}
result <- runCPie(
    counts,
    assignments,
    image = tissue,
    spotCoordinates = spots
)
result
```
